### ISSUE TITLE
Build: simplify `_config` attribute data modelling

### ISSUE THREAD

**Issue by humitos** [readthedocs/readthedocs.org#10586]
We are currently using a `models.JSONField` for the `Build._config` attribute. This field can be used in two different ways:

1. store the whole YAML configuration file
2. save a build ID where the full/whole YAML configuration file is stored

This is basically to avoid duplicating the data on our database. However, it only avoids duplicated config on the same projects and only for consecutive builds.

I found this approach pretty confusing when exploring `Build` objects because we have to do extra operations to find out the real YAML used for that build (see for example the `config @property` at https://github.com/readthedocs/readthedocs.org/blob/12cbb3d67efec73bdcd22b710349ebe10a3ec70f/readthedocs/builds/models.py#L313-L331)

Also, it makes hard to answer questions like "What are the projects using `build.os: ubuntu-22.04`?" in an easy way and many other similar questions that require querying the config file.

### New approach proposal

Due to these limitations and complications, I want to propose creating a new model called `BuildConfig` that's used as a foreign key from the `Build` model:

```python
class Build(models.Model):
    # ...
    readthedocs_yaml_data = models.ForeignKey(
        "BuildConfig"
        null=True,
        blank=True,
    )
    # ...
```

Then the new model would be:

```python
class BuildConfig(TimeStampedModel):
    # ...
    data = models.JSONField(unique=True)
    # ...
```

### Benefits

- the "avoid duplicated" data is solved automatically by the database at a table level among _all the projects_, instead of per-project (e.g. `BuildConfig.objects.get_or_create(data=...)`)
- getting the YAML for a particular build is going to be _pretty fast_ since we have direct access to it
- we can reduce the number of queries performed by using `.select_related("readthedocs_yaml_data")`
- answer interesting/complex questions pretty easily using Django ORM (e.g. `BuildConfig.objects.filter(data__build__os="ubuntu-22.04").projects.count()` to answer my previous example question)
- remove the requirement of a helper `@property` to get the YAML for a particular build
- remove the requirement for a `@config.setter` helper
- remove the requirement of having extra logic at `Build.save`
- remove the need of using `self._config_changed`
- remove having two different meanings for the same field
- allows us to quickly show badges on "Build details page" as Anthony suggested for ext-theme

Summarizing, this approach keeps the same features but simplifies the application and modelling and gives us more features to analyze in an easier way platform usage and make better decisions based on them.

### Rollout proposal

1. add a `Build.readthedocs_yaml_data` field and create the new `BuildConfig` model
2. start saving both fields `Build.readthedocs_yaml_data` and `Build._config` while doing the migration
3. make a data migration to convert `Build._config` into `BuildConfig` and link `Build.readthedocs_yaml_data` to them
4. make the application to use the new `Build.readthedocs_yaml_data` field
5. remove the old code that uses `Build._config` field

**Comment by humitos** [comment]
Another important benefit that I'm seeing here about making querying our database easier: while working on the deprecation of config file and `build.image` I had to make multiple queries and iterate over all the builds returned to get the _real_ value of `Build._config` https://github.com/readthedocs/readthedocs.org/blob/1b74999add86d6a45f4c6f6149dcb21d5ef2322b/readthedocs/projects/tasks/utils.py#L235-L251

Instead, we would be able to do all that in just one query:

```python
Project.objects.filter(
  builds__version__slug=F("default_version"),
  builds__date__gt=timezone.now() - timezone.timedelta(days=365),
  builds__success=True,
  builds___config__build__image="readthedocs/build:latest",
)
.distinct()
```

That query will give me "projects building with `build.image: readthedocs/build:latest` on their default version with a successful build in the last year", which is a lot easier to read, maintain and communicate to the rest of the team.

Note that with the current structure, we won't get the accurate data because those builds where the config file is identical to their previous build will have `_config` defined as `{"__config": 12345678}"`

**Comment by ericholscher** [comment]
Yeah, it seems like our implementation is just a hacky way of doing a foreignkey, without any of the nice benefits. This seems like an obvious thing to move forward on, I don't really see a downside.
